lib/kube/proxy/forwarder.go

```diff
@@ -2148,3 +2312,358 @@ func (f *Forwarder) removeKubeDetails(name string) {
+	}
+	// Forwards the request to the target cluster.
+	sess.forwarder.ServeHTTP(rw, req)
+	// Once the request terminates, close the watcher and waits for resources
+	// cleanup.
+	err = rw.Close()
+	return rw.Status(), trace.Wrap(err)
+}
+
+// deletePodsCollection calls checks the listPodsList
+func (f *Forwarder) deletePodsCollection(ctx *authContext, w http.ResponseWriter, req *http.Request, p httprouter.Params) (resp interface{}, err error) {
+	sess, err := f.newClusterSession(*ctx)
+	if err != nil {
+		// This error goes to kubernetes client and is not visible in the logs
+		// of the teleport server if not logged here.
+		f.log.Errorf("Failed to create cluster session: %v.", err)
+		return nil, trace.Wrap(err)
+	}
+
+	sess.upgradeToHTTP2 = true
+	sess.forwarder, err = f.makeSessionForwarder(sess)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := f.setupForwardingHeaders(sess, req); err != nil {
+		// This error goes to kubernetes client and is not visible in the logs
+		// of the teleport server if not logged here.
+		f.log.Errorf("Failed to set up forwarding headers: %v.", err)
+		return nil, trace.Wrap(err)
+	}
+	// status holds the returned response code.
+	var status int
+	// Check if the target Kubernetes cluster is not served by the current service.
+	// If it's the case, forward the request to the target Kube Service where the
+	// filtering logic will be applied.
+	if !f.isLocalKubeCluster(sess) {
+		rw := httplib.NewResponseStatusRecorder(w)
+		sess.forwarder.ServeHTTP(rw, req)
+		status = rw.Status()
+	} else {
+		memoryRW := responsewriters.NewMemoryResponseWriter()
+		listReq := req.Clone(req.Context())
+		// reset body and method since list does not need the body response.
+		listReq.Body = nil
+		listReq.Method = http.MethodGet
+		_, err = f.listPods(ctx, memoryRW, listReq, p)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		status, err = f.handleDeleteCollectionReq(req, &sess.authContext, memoryRW, w)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	if sess.noAuditEvents {
+		return nil, nil
+	}
+	// Emit audit event.
+	event := &apievents.KubeRequest{
+		Metadata: apievents.Metadata{
+			Type: events.KubeRequestEvent,
+			Code: events.KubeRequestCode,
+		},
+		UserMetadata: ctx.eventUserMeta(),
+		ConnectionMetadata: apievents.ConnectionMetadata{
+			RemoteAddr: req.RemoteAddr,
+			LocalAddr:  sess.kubeAddress,
+			Protocol:   events.EventProtocolKube,
+		},
+		ServerMetadata: apievents.ServerMetadata{
+			ServerID:        f.cfg.HostID,
+			ServerNamespace: f.cfg.Namespace,
+		},
+		RequestPath:               req.URL.Path,
+		Verb:                      req.Method,
+		ResponseCode:              int32(status),
+		KubernetesClusterMetadata: ctx.eventClusterMeta(),
+	}
+	r := parseResourcePath(req.URL.Path)
+	if r.skipEvent {
+		return nil, nil
+	}
+	r.populateEvent(event)
+	if err := f.cfg.AuthClient.EmitAuditEvent(f.ctx, event); err != nil {
```

Comment:
Similar to above, there seems to be a common boilerplate for audit logs. Refactor?